### DISCUSSION TITLE
Async Database Access: PostgreSQL vs. MSSQL

### DISCUSSION THREAD

**Discussion by jordiarellano1996** [sqlalchemy/sqlalchemy#10561] [Usage Questions] [answered]
Why does PostgreSQL have a DBAPI that leverages Python's asynchronous capabilities, while for Microsoft SQL Server, we may need to use a thread wrapper like aioodbc to prevent Python thread blocking? What are the key differences between PostgreSQL and MSSQL in this context? 
Thanks to the community for helping me understand, as I'm discussing this at work and need some insights.
J.

**Comment by CaselIT** [comment]
Hi,

Don't know. From what I can see Postgresql is the odd one here: oracle has no async library, sqlite has one that uses threads, mysql has an async dbapi but it's not very popular or one that uses threads.

Most likely python projects (at least the async ones)  mainly use/care about postgresql as db of choice

**Comment by zzzeek** [comment, marked as answer]
PostgreSQL is unique from the other major database servers in that it has a [fully native asynchronous command protocol](https://www.postgresql.org/docs/current/libpq-async.html).   To my knowledge there is no equivalent feature in MySQL's protocol, or odbc/ TDS, or Oracle.   this is why async drivers for PostgreSQL are common and native while for the other backends they need to use threadpools.
